### PR TITLE
Isolate the code bundled in the PHARs

### PR DESCRIPTION
**Pull request by theofidry**

### CODE REVIEW

**Comment by theofidry** [comment]
@bobthecow seems to be working but I'm a bit concerned at the lack of e2e tests. That said I'm not sure I'm willing to dig too deep into that either 😅

**Comment by bobthecow** [comment]
Wow. Is it really 400x slower to build the phars this way?

**Comment by theofidry** [comment]
For sure, all files are parsed with PHP-Parser and the resulting AST tree iterated over for analysis/change...

File: bin/build-stub [hunk 1]
```diff
@@ -7,7 +7,7 @@ $license = str_replace("\n", "\n * ", trim($license));
 
 $autoload = <<<'EOS'
     Phar::mapPhar('psysh.phar');
-    require 'phar://psysh.phar/.box/check_requirements.php';
+    require 'phar://psysh.phar/.box/bin/check-requirements.php';
```
Comment (bobthecow) [inline bin/build-stub]: was there a box version change that this happened in? should we bump our `composer.json` to reflect it?

File: scoper.inc.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+return [
+    'whitelist' => [
+        'Psy\Shell',
+        'Psy\bin',
```
Comment (bobthecow) [inline scoper.inc.php]: if this means "things that shouldn't be renamed", then `Psy\sh`, `Psy\debug`, and `Psy\info` belong in this list as well.

File: bin/build-stub (hunk 1 above)
Comment (theofidry) [reply bin/build-stub]: Yep there was one, but `.lock` is commited for Box so it's ok

**Comment by theofidry** [comment]
>Knowing HOA, it's going to need something; possibly completely excluded as well? Since it's only used in the *-compat builds

Arg, right, I had to add a couple of patches for it in Box itself... will do and test it here as well

>I'm kind of leaning towards excluding all of Psy namespace from php-scoper.
>[...] The biggest worry is with third-party dependencies, not with psy/psysh itself

That's fair enough.

Will update shortly

File: bin/build-stub (hunk 1 above)
Comment (bobthecow) [reply bin/build-stub]: Until the next time the lock file is updated ;)

Comment (theofidry) [reply bin/build-stub]: Fair

**Comment by theofidry** [comment]
Updated

**Comment by bobthecow** [comment]
my first couple of builds were a little weird, but `make clean` cleaned it up and they seem to be building fine now. which is a smell that we missed something in our makefile, but i can't figure out what it would be :( 

I found a couple of references that didn't get scoped properly.

### in all the phars:
```
                $messages = \array_map(['Symfony\\Component\\Console\\Formatter\\OutputFormatter', 'escape'], $messages);
```
```
                    switch ($ns = \substr($name, 0, $len)) {
                        case 'Symfony\\Bridge\\':
                        case 'Symfony\\Bundle\\':
                        case 'Symfony\\Component\\':
                            $ns = 'Symfony\\';
                            $len = \strlen($ns);
                            break;
                    }
```
```
    private function classNameFromNodeType(string $nodeType) : string
    {
        $className = 'PhpParser\\Node\\' . \strtr($nodeType, '_', '\\');
        if (\class_exists($className)) {
            return $className;
        }
```

### in the 5.4 phars:
```
    protected function readComment()
    {
        $className = $this->reader->getAttribute('isDocComment') === 'true' ? 'PhpParser\\Comment\\Doc' : 'PhpParser\\Comment';
        return new $className($this->reader->readString(), $this->parseInt($this->reader->getAttribute('line')));
    }
```
```
                    if (\defined($name = 'PhpParser\\Parser\\Tokens::' . $name)) {
```
```
    public static function autoload($class)
    {
        if (0 === \strpos($class, 'PhpParser\\')) {
            $fileName = __DIR__ . \strtr(\substr($class, 9), '\\', '/') . '.php';
            if (\file_exists($fileName)) {
                require $fileName;
            }
        }
    }
```
```
    public static $defaultCasters = array('_HumbugBoxa0c4b7107ccc\\Symfony\\Component\\VarDumper\\Caster\\CutStub' => 'Symfony\\Component\\VarDumper\\Caster\\StubCaster::castStub', '_HumbugBoxa0c4b7107ccc\\Symfony\\Component\\VarDumper\\Caster\\CutArrayStub' => 'Symfony\\Component\\VarDumper\\Caster\\StubCaster::castCutArray', '_HumbugBoxa0c4b7107ccc\\Symfony\\Component\\VarDumper\\Caster\\ConstStub' => 'Symfony\\Component\\VarDumper\\Caster\\StubCaster::castStub', '_HumbugBoxa0c4b7107ccc\\Symfony\\Component\\VarDumper\\Caster\\EnumStub' => 'Symfony\\Component\\VarDumper\\Caster\\StubCaster::castEnum', 'Closure' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castClosure', 'Generator' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castGenerator', 'ReflectionType' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castType', 'ReflectionGenerator' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castReflectionGenerator', 'ReflectionClass' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castClass', 'ReflectionFunctionAbstract' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castFunctionAbstract', 'ReflectionMethod' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castMethod', 'ReflectionParameter' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castParameter', 'ReflectionProperty' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castProperty', 'ReflectionExtension' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castExtension', 'ReflectionZendExtension' => 'Symfony\\Component\\VarDumper\\Caster\\ReflectionCaster::castZendExtension', '_HumbugBoxa0c4b7107ccc\\Doctrine\\Common\\Persistence\\ObjectManager' => 'Symfony\\Component\\VarDumper\\Caster\\StubCaster::cutInternals', '_HumbugBoxa0c4b7107ccc\\Doctrine\\Common\\Proxy\\Proxy' => 'Symfony\\Component\\VarDumper\\Caster\\DoctrineCaster::castCommonProxy', '_HumbugBoxa0c4b7107ccc\\Doctrine\\ORM\\Proxy\\Proxy' => 'Symfony\\Component\\VarDumper\\Caster\\DoctrineCaster::castOrmProxy', '_HumbugBoxa0c4b7107ccc\\Doctrine\\ORM\\PersistentCollection' => 'Symfony\\Component\\VarDumper\\Caster\\DoctrineCaster::castPersistentCollection', 'DOMException' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castException', 'DOMStringList' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castLength', 'DOMNameList' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castLength', 'DOMImplementation' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castImplementation', 'DOMImplementationList' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castLength', 'DOMNode' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castNode', 'DOMNameSpaceNode' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castNameSpaceNode', 'DOMDocument' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castDocument', 'DOMNodeList' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castLength', 'DOMNamedNodeMap' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castLength', 'DOMCharacterData' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castCharacterData', 'DOMAttr' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castAttr', 'DOMElement' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castElement', 'DOMText' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castText', 'DOMTypeinfo' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castTypeinfo', 'DOMDomError' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castDomError', 'DOMLocator' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castLocator', 'DOMDocumentType' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castDocumentType', 'DOMNotation' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castNotation', 'DOMEntity' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castEntity', 'DOMProcessingInstruction' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castProcessingInstruction', 'DOMXPath' => 'Symfony\\Component\\VarDumper\\Caster\\DOMCaster::castXPath', 'ErrorException' => 'Symfony\\Component\\VarDumper\\Caster\\ExceptionCaster::castErrorException', 'Exception' => 'Symfony\\Component\\VarDumper\\Caster\\ExceptionCaster::castException', 'Error' => 'Symfony\\Component\\VarDumper\\Caster\\ExceptionCaster::castError', '_HumbugBoxa0c4b7107ccc\\Symfony\\Component\\DependencyInjection\\ContainerInterface' => 'Symfony\\Component\\VarDumper\\Caster\\StubCaster::cutInternals', '_HumbugBoxa0c4b7107ccc\\Symfony\\Component\\VarDumper\\Exception\\ThrowingCasterException' => 'Symfony\\Component\\VarDumper\\Caster\\ExceptionCaster::castThrowingCasterException', '_HumbugBoxa0c4b7107ccc\\Symfony\\Component\\VarDumper\\Caster\\TraceStub' => 'Symfony\\Component\\VarDumper\\Caster\\ExceptionCaster::castTraceStub', '_HumbugBoxa0c4b7107ccc\\Symfony\\Component\\VarDumper\\Caster\\FrameStub' => 'Symfony\\Component\\VarDumper\\Caster\\ExceptionCaster::castFrameStub', 'PHPUnit_Framework_MockObject_MockObject' => 'Symfony\\Component\\VarDumper\\Caster\\StubCaster::cutInternals', '_HumbugBoxa0c4b7107ccc\\Prophecy\\Prophecy\\ProphecySubjectInterface' => 'Symfony\\Component\\VarDumper\\Caster\\StubCaster::cutInternals', '_HumbugBoxa0c4b7107ccc\\Mockery\\MockInterface' => 'Symfony\\Component\\VarDumper\\Caster\\StubCaster::cutInternals', 'PDO' => 'Symfony\\Component\\VarDumper\\Caster\\PdoCaster::castPdo', 'PDOStatement' => 'Symfony\\Component\\VarDumper\\Caster\\PdoCaster::castPdoStatement', 'AMQPConnection' => 'Symfony\\Component\\VarDumper\\Caster\\AmqpCaster::castConnection', 'AMQPChannel' => 'Symfony\\Component\\VarDumper\\Caster\\AmqpCaster::castChannel', 'AMQPQueue' => 'Symfony\\Component\\VarDumper\\Caster\\AmqpCaster::castQueue', 'AMQPExchange' => 'Symfony\\Component\\VarDumper\\Caster\\AmqpCaster::castExchange', 'AMQPEnvelope' => 'Symfony\\Component\\VarDumper\\Caster\\AmqpCaster::castEnvelope', 'ArrayObject' => 'Symfony\\Component\\VarDumper\\Caster\\SplCaster::castArrayObject', 'ArrayIterator' => 'Symfony\\Component\\VarDumper\\Caster\\SplCaster::castArrayIterator', 'SplDoublyLinkedList' => 'Symfony\\Component\\VarDumper\\Caster\\SplCaster::castDoublyLinkedList', 'SplFileInfo' => 'Symfony\\Component\\VarDumper\\Caster\\SplCaster::castFileInfo', 'SplFileObject' => 'Symfony\\Component\\VarDumper\\Caster\\SplCaster::castFileObject', 'SplFixedArray' => 'Symfony\\Component\\VarDumper\\Caster\\SplCaster::castFixedArray', 'SplHeap' => 'Symfony\\Component\\VarDumper\\Caster\\SplCaster::castHeap', 'SplObjectStorage' => 'Symfony\\Component\\VarDumper\\Caster\\SplCaster::castObjectStorage', 'SplPriorityQueue' => 'Symfony\\Component\\VarDumper\\Caster\\SplCaster::castHeap', 'OuterIterator' => 'Symfony\\Component\\VarDumper\\Caster\\SplCaster::castOuterIterator', 'MongoCursorInterface' => 'Symfony\\Component\\VarDumper\\Caster\\MongoCaster::castCursor', ':curl' => 'Symfony\\Component\\VarDumper\\Caster\\ResourceCaster::castCurl', ':dba' => 'Symfony\\Component\\VarDumper\\Caster\\ResourceCaster::castDba', ':dba persistent' => 'Symfony\\Component\\VarDumper\\Caster\\ResourceCaster::castDba', ':gd' => 'Symfony\\Component\\VarDumper\\Caster\\ResourceCaster::castGd', ':mysql link' => 'Symfony\\Component\\VarDumper\\Caster\\ResourceCaster::castMysqlLink', ':pgsql large object' => 'Symfony\\Component\\VarDumper\\Caster\\PgSqlCaster::castLargeObject', ':pgsql link' => 'Symfony\\Component\\VarDumper\\Caster\\PgSqlCaster::castLink', ':pgsql link persistent' => 'Symfony\\Component\\VarDumper\\Caster\\PgSqlCaster::castLink', ':pgsql result' => 'Symfony\\Component\\VarDumper\\Caster\\PgSqlCaster::castResult', ':process' => 'Symfony\\Component\\VarDumper\\Caster\\ResourceCaster::castProcess', ':stream' => 'Symfony\\Component\\VarDumper\\Caster\\ResourceCaster::castStream', ':persistent stream' => 'Symfony\\Component\\VarDumper\\Caster\\ResourceCaster::castStream', ':stream-context' => 'Symfony\\Component\\VarDumper\\Caster\\ResourceCaster::castStreamContext', ':xml' => 'Symfony\\Component\\VarDumper\\Caster\\XmlResourceCaster::castXml');
```
```
        $a = $this->callCaster('Symfony\\Component\\VarDumper\\Caster\\Caster::castObject', $obj, $classInfo[0], null, $isNested);
```
```
        return \call_user_func_array(array('Symfony\\Component\\Debug\\Exception\\FlattenException', $method), $args);
```

### in the compat phars:
```
\_HumbugBoxb886a4cd1570\Hoa\Consistency::registerShutdownFunction(xcallable('Hoa\\Console\\Mouse::untrack'));
```
```
\_HumbugBoxb886a4cd1570\Hoa\Consistency::registerShutdownFunction(xcallable('Hoa\\Console\\Console::restoreInteraction'));
```
```
        if ('Hoa\\Core' === \substr($classname, 0, 8) && \false !== ($pos = \strpos($classname, '\\', 10)) && 'Bin\\' !== \substr($classname, 9, 4)) {
```
```
        if ($this->_owner == 'Hoa\\Core\\Core') {
            $class = 'HoaCoreCore';
            $this->_parameters = $parameters;
            $path = $this->zFormat($parameters['protocol.Data/Etc/Configuration']) . '.Cache' . DS . 'HoaCoreCore.php';
        } else {
```

**Comment by bobthecow** [comment]
this sort of thing could be a linter / warning that php scoper emits. i found all the prefixes in my composer classmap, then searched the built phar for them:

```
[^\\](XdgBaseDir|Symfony|Psr|PhpParser|JakubOnderka|Hoa)\\
```

**Comment by theofidry** [comment]
I try to get PhpScoper be able to scope as much as possible but there is
things a bit hard to guess so not sure how the linting would work :/

**Comment by theofidry** [comment]
Got the same feeling about them makzfile but couldn’t figure out why yet

**Comment by theofidry** [comment]
@bobthecow updated

**Comment by bobthecow** [comment]
thank you. it's looking good!

i hacked the tests up a bit to get them to run against the scoped phar rather than the composer vendor directory:

<details>
<summary>It was a surprisingly small diff 🙂</summary>

```diff
diff --git a/autoload.php b/autoload.php
new file mode 100644
index 0000000..d33678e
--- /dev/null
+++ b/autoload.php
@@ -0,0 +1,13 @@
+<?php
+
+/*
+ * This file is part of Psy Shell.
+ *
+ * (c) 2012-2018 Justin Hileman
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+require __DIR__ . '/build/psysh/psysh';
+require __DIR__ . '/vendor/autoload.php';
diff --git a/bin/psysh b/bin/psysh
index 7dbd203..5c7b915 100755
--- a/bin/psysh
+++ b/bin/psysh
@@ -13,6 +13,8 @@
 // Try to find an autoloader for a local psysh version.
 // We'll wrap this whole mess in a Closure so it doesn't leak any globals.
 call_user_func(function () {
+    return;
+
     $cwd = null;
 
     // Find the cwd arg (if present)
diff --git a/composer.json b/composer.json
index 4588ba6..90c7432 100644
--- a/composer.json
+++ b/composer.json
@@ -23,7 +23,6 @@
         "jakub-onderka/php-console-highlighter": "0.3.*|0.4.*"
     },
     "require-dev": {
-        "phpunit/phpunit": "~4.8.35|~5.0|~6.0|~7.0",
         "hoa/console": "~2.15|~3.16",
         "bamarni/composer-bin-plugin": "^1.2"
     },
@@ -34,12 +33,6 @@
         "ext-pdo-sqlite": "The doc command requires SQLite to work.",
         "hoa/console": "A pure PHP readline implementation. You'll want this if your PHP install doesn't already support readline or libedit."
     },
-    "autoload": {
-        "files": ["src/functions.php"],
-        "psr-4": {
-            "Psy\\": "src/"
-        }
-    },
     "autoload-dev": {
         "psr-4": {
             "Psy\\Test\\": "test/"
diff --git a/phpunit.xml.dist b/phpunit.xml.dist
index 3534fca..e82a430 100644
--- a/phpunit.xml.dist
+++ b/phpunit.xml.dist
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<phpunit backupGlobals="false" colors="true" bootstrap="vendor/autoload.php">
+<phpunit backupGlobals="false" colors="true" bootstrap="autoload.php">
 	<testsuite name="PsySH">
 		<directory suffix="Test.php">./test</directory>
 	</testsuite>
diff --git a/test/CodeCleaner/CodeCleanerTestCase.php b/test/CodeCleaner/CodeCleanerTestCase.php
index 6962c11..c86ca99 100644
--- a/test/CodeCleaner/CodeCleanerTestCase.php
+++ b/test/CodeCleaner/CodeCleanerTestCase.php
@@ -11,7 +11,7 @@
 
 namespace Psy\Test\CodeCleaner;
 
-use PhpParser\NodeTraverser;
+use _HumbugBox29e42959e764\PhpParser\NodeTraverser;
 use Psy\CodeCleaner\CodeCleanerPass;
 use Psy\Test\ParserTestCase;
 
diff --git a/test/CodeCleaner/NoReturnValueTest.php b/test/CodeCleaner/NoReturnValueTest.php
index 11e9d01..21b6fc1 100644
--- a/test/CodeCleaner/NoReturnValueTest.php
+++ b/test/CodeCleaner/NoReturnValueTest.php
@@ -11,7 +11,7 @@
 
 namespace Psy\Test\CodeCleaner;
 
-use PhpParser\Node\Stmt\Expression;
+use _HumbugBox29e42959e764\PhpParser\Node\Stmt\Expression;
 use Psy\CodeCleaner\NoReturnValue;
 use Psy\Test\ParserTestCase;
 
diff --git a/test/Command/ExitCommandTest.php b/test/Command/ExitCommandTest.php
index 45626be..a174bf6 100644
--- a/test/Command/ExitCommandTest.php
+++ b/test/Command/ExitCommandTest.php
@@ -11,8 +11,8 @@
 
 namespace Symfony\Component\Console\Tests\Command;
 
+use _HumbugBox29e42959e764\Symfony\Component\Console\Tester\CommandTester;
 use Psy\Command\ExitCommand;
-use Symfony\Component\Console\Tester\CommandTester;
 
 class ExitCommandTest extends \PHPUnit\Framework\TestCase
 {
diff --git a/test/Command/ThrowUpCommandTest.php b/test/Command/ThrowUpCommandTest.php
index 511c580..3ad3d28 100644
--- a/test/Command/ThrowUpCommandTest.php
+++ b/test/Command/ThrowUpCommandTest.php
@@ -11,9 +11,9 @@
 
 namespace Symfony\Component\Console\Tests\Command;
 
+use _HumbugBox29e42959e764\Symfony\Component\Console\Tester\CommandTester;
 use Psy\Command\ThrowUpCommand;
 use Psy\Shell;
-use Symfony\Component\Console\Tester\CommandTester;
 
 class ThrowUpCommandTest extends \PHPUnit\Framework\TestCase
 {
@@ -76,7 +76,7 @@ class ThrowUpCommandTest extends \PHPUnit\Framework\TestCase
     }
 
     /**
-     * @expectedException \PhpParser\Error
+     * @expectedException \_HumbugBox29e42959e764\PhpParser\Error
      * @expectedExceptionMessage Syntax error, unexpected ')' on line 1
      */
     public function testParseErrorThrowsException()
diff --git a/test/Command/TimeitCommand/TimeitVisitorTest.php b/test/Command/TimeitCommand/TimeitVisitorTest.php
index 1f31735..1918070 100644
--- a/test/Command/TimeitCommand/TimeitVisitorTest.php
+++ b/test/Command/TimeitCommand/TimeitVisitorTest.php
@@ -11,7 +11,7 @@
 
 namespace Psy\Test\Command\TimeitCommand;
 
-use PhpParser\NodeTraverser;
+use _HumbugBox29e42959e764\PhpParser\NodeTraverser;
 use Psy\Command\TimeitCommand\TimeitVisitor;
 use Psy\Test\ParserTestCase;
 
diff --git a/test/ConfigurationTest.php b/test/ConfigurationTest.php
index b7b6c0c..dab1e1e 100644
--- a/test/ConfigurationTest.php
+++ b/test/ConfigurationTest.php
@@ -11,11 +11,11 @@
 
 namespace Psy\Test;
 
+use _HumbugBox29e42959e764\Symfony\Component\Console\Output\ConsoleOutput;
 use Psy\CodeCleaner;
 use Psy\Configuration;
 use Psy\Output\PassthruPager;
 use Psy\VersionUpdater\GitHubChecker;
-use Symfony\Component\Console\Output\ConsoleOutput;
 
 class ConfigurationTest extends \PHPUnit\Framework\TestCase
 {
diff --git a/test/Exception/ParseErrorExceptionTest.php b/test/Exception/ParseErrorExceptionTest.php
index 446b938..c99c441 100644
--- a/test/Exception/ParseErrorExceptionTest.php
+++ b/test/Exception/ParseErrorExceptionTest.php
@@ -20,7 +20,7 @@ class ParseErrorExceptionTest extends \PHPUnit\Framework\TestCase
         $e = new ParseErrorException();
 
         $this->assertInstanceOf('Psy\Exception\Exception', $e);
-        $this->assertInstanceOf('PhpParser\Error', $e);
+        $this->assertInstanceOf('\_HumbugBox29e42959e764\PhpParser\Error', $e);
         $this->assertInstanceOf('Psy\Exception\ParseErrorException', $e);
     }
 
@@ -34,7 +34,7 @@ class ParseErrorExceptionTest extends \PHPUnit\Framework\TestCase
 
     public function testConstructFromParseError()
     {
-        $e = ParseErrorException::fromParseError(new \PhpParser\Error('{msg}'));
+        $e = ParseErrorException::fromParseError(new \_HumbugBox29e42959e764\PhpParser\Error('{msg}'));
 
         $this->assertContains('{msg}', $e->getRawMessage());
         $this->assertContains('PHP Parse error:', $e->getMessage());
diff --git a/test/Formatter/SignatureFormatterTest.php b/test/Formatter/SignatureFormatterTest.php
index 096e6ae..3ee5d5f 100644
--- a/test/Formatter/SignatureFormatterTest.php
+++ b/test/Formatter/SignatureFormatterTest.php
@@ -54,8 +54,8 @@ class SignatureFormatterTest extends \PHPUnit\Framework\TestCase
             [
                 new \ReflectionClass('Psy\CodeCleaner\CodeCleanerPass'),
                 'abstract class Psy\CodeCleaner\CodeCleanerPass '
-                . 'extends PhpParser\NodeVisitorAbstract '
-                . 'implements PhpParser\NodeVisitor',
+                . 'extends _HumbugBox29e42959e764\PhpParser\NodeVisitorAbstract '
+                . 'implements _HumbugBox29e42959e764\PhpParser\NodeVisitor',
             ],
             [
                 new \ReflectionFunction('array_chunk'),
diff --git a/test/Input/CodeArgumentTest.php b/test/Input/CodeArgumentTest.php
index c396152..5579021 100644
--- a/test/Input/CodeArgumentTest.php
+++ b/test/Input/CodeArgumentTest.php
@@ -11,8 +11,8 @@
 
 namespace Psy\Test\Input;
 
+use _HumbugBox29e42959e764\Symfony\Component\Console\Input\InputArgument;
 use Psy\Input\CodeArgument;
-use Symfony\Component\Console\Input\InputArgument;
 
 class CodeArgumentTest extends \PHPUnit\Framework\TestCase
 {
diff --git a/test/Input/FilterOptionsTest.php b/test/Input/FilterOptionsTest.php
index 7ed8919..e854094 100644
--- a/test/Input/FilterOptionsTest.php
+++ b/test/Input/FilterOptionsTest.php
@@ -11,9 +11,9 @@
 
 namespace Psy\Test\Input;
 
+use _HumbugBox29e42959e764\Symfony\Component\Console\Input\InputDefinition;
+use _HumbugBox29e42959e764\Symfony\Component\Console\Input\StringInput;
 use Psy\Input\FilterOptions;
-use Symfony\Component\Console\Input\InputDefinition;
-use Symfony\Component\Console\Input\StringInput;
 
 class FilterOptionsTest extends \PHPUnit\Framework\TestCase
 {
diff --git a/test/Input/ShellInputTest.php b/test/Input/ShellInputTest.php
index f19d126..0cb49fa 100644
--- a/test/Input/ShellInputTest.php
+++ b/test/Input/ShellInputTest.php
@@ -11,11 +11,11 @@
 
 namespace Psy\Test\Input;
 
+use _HumbugBox29e42959e764\Symfony\Component\Console\Input\InputArgument;
+use _HumbugBox29e42959e764\Symfony\Component\Console\Input\InputDefinition;
+use _HumbugBox29e42959e764\Symfony\Component\Console\Input\InputOption;
 use Psy\Input\CodeArgument;
 use Psy\Input\ShellInput;
-use Symfony\Component\Console\Input\InputArgument;
-use Symfony\Component\Console\Input\InputDefinition;
-use Symfony\Component\Console\Input\InputOption;
 
 class ShellInputTest extends \PHPUnit\Framework\TestCase
 {
diff --git a/test/ParserTestCase.php b/test/ParserTestCase.php
index a0740b6..96708c5 100644
--- a/test/ParserTestCase.php
+++ b/test/ParserTestCase.php
@@ -11,7 +11,7 @@
 
 namespace Psy\Test;
 
-use PhpParser\PrettyPrinter\Standard as Printer;
+use _HumbugBox29e42959e764\PhpParser\PrettyPrinter\Standard as Printer;
 use Psy\Exception\ParseErrorException;
 use Psy\ParserFactory;
 
@@ -33,7 +33,7 @@ class ParserTestCase extends \PHPUnit\Framework\TestCase
         $code = $prefix . $code;
         try {
             return $this->getParser()->parse($code);
-        } catch (\PhpParser\Error $e) {
+        } catch (\_HumbugBox29e42959e764\PhpParser\Error $e) {
             if (!$this->parseErrorIsEOF($e)) {
                 throw ParseErrorException::fromParseError($e);
             }
@@ -41,7 +41,7 @@ class ParserTestCase extends \PHPUnit\Framework\TestCase
             try {
                 // Unexpected EOF, try again with an implicit semicolon
                 return $this->getParser()->parse($code . ';');
-            } catch (\PhpParser\Error $e) {
+            } catch (\_HumbugBox29e42959e764\PhpParser\Error $e) {
                 return false;
             }
         }
@@ -88,7 +88,7 @@ class ParserTestCase extends \PHPUnit\Framework\TestCase
         return $this->printer;
     }
 
-    private function parseErrorIsEOF(\PhpParser\Error $e)
+    private function parseErrorIsEOF(\_HumbugBox29e42959e764\PhpParser\Error $e)
     {
         $msg = $e->getRawMessage();
 
diff --git a/test/ShellTest.php b/test/ShellTest.php
index 809af02..2be6236 100644
--- a/test/ShellTest.php
+++ b/test/ShellTest.php
@@ -11,12 +11,12 @@
 
 namespace Psy\Test;
 
+use _HumbugBox29e42959e764\Symfony\Component\Console\Output\StreamOutput;
 use Psy\Configuration;
 use Psy\Exception\ErrorException;
 use Psy\Exception\ParseErrorException;
 use Psy\Shell;
 use Psy\TabCompletion\Matcher\ClassMethodsMatcher;
-use Symfony\Component\Console\Output\StreamOutput;
 
 class ShellTest extends \PHPUnit\Framework\TestCase
 {
@@ -218,7 +218,7 @@ class ShellTest extends \PHPUnit\Framework\TestCase
     {
         $shell = new Shell($this->getConfig());
 
-        $this->assertInstanceOf('Symfony\Component\Console\Application', $shell);
+        $this->assertInstanceOf('_HumbugBox29e42959e764\Symfony\Component\Console\Application', $shell);
         $this->assertContains(Shell::VERSION, $shell->getVersion());
         $this->assertContains(PHP_VERSION, $shell->getVersion());
         $this->assertContains(PHP_SAPI, $shell->getVersion());
diff --git a/test/Sudo/SudoVisitorTest.php b/test/Sudo/SudoVisitorTest.php
index 1f2042c..9db150b 100644
--- a/test/Sudo/SudoVisitorTest.php
+++ b/test/Sudo/SudoVisitorTest.php
@@ -11,7 +11,7 @@
 
 namespace Psy\Test\Sudo;
 
-use PhpParser\NodeTraverser;
+use _HumbugBox29e42959e764\PhpParser\NodeTraverser;
 use Psy\Sudo\SudoVisitor;
 use Psy\Test\ParserTestCase;
 
diff --git a/vendor-bin/phpunit/composer.json b/vendor-bin/phpunit/composer.json
new file mode 100644
index 0000000..526dba5
--- /dev/null
+++ b/vendor-bin/phpunit/composer.json
@@ -0,0 +1,5 @@
+{
+    "require": {
+        "phpunit/phpunit": "^6"
+    }
+}
```
</details>

<br/>

The only failure was with `Hoa\Console`, which isn't super surprising because it's not included in the non-compat phar, so it would be looking for it in the vendor directory anyway:

```
1) Psy\Test\Readline\HoaConsoleTest::testHistory
Error: Class '_HumbugBox29e42959e764\Hoa\Console\Readline\Readline' not found

phar:///Users/justin/Projects/public/psysh/build/psysh/psysh/src/Readline/HoaConsole.php:16
/Users/justin/Projects/public/psysh/test/Readline/HoaConsoleTest.php:20
```

**Comment by theofidry** [comment]
@bobthecow cool, how can I reproduce it to fix the Readline thingy?

**Comment by bobthecow** [comment]
Oh, sorry that wasn’t super clear. This is a failure of the tests, not a test failure. 

I ran everything against the normal (non-compat) phar, which meant it was missing Hoa Console entirely. So there’s no way that test could have passed.

**Comment by theofidry** [comment]
Oh I see. I think the failing build in the CI needs a restart and it should be green. Is there anything else missing?

**Comment by bobthecow** [comment]
I've restarted it three time :-/

**Comment by theofidry** [comment]
maybe we cna add `--no-parallel`

**Comment by theofidry** [comment]
It's gonna be super slow though :(

**Comment by theofidry** [comment]
Nice! I hope not too many issues will arise with the scoping 🤞

Any ETA about when this is gonna be released?

**Comment by bobthecow** [comment]
It's going to have to be v0.10.0 because it breaks backwards compatibility (ever so slightly) if you're using the phar to load code rather than to just run psysh. I'll have to think about ETA a bit because there are some other breaking changes I was considering :)

**Comment by bobthecow** [comment]
@theofidry this has made builds super flaky and slow. we're going to need to do something differently :(

**Comment by theofidry** [comment]
The issue with that unstable build?

**Comment by theofidry** [comment]
Regarding the error:

>[ERROR] Amp\Parallel\Sync\ChannelException: The channel closed unexpectedly.   

I think we should report it. It's happening often enough and without a clear reason so it might be a legit  bug. It is also worth noting I never had it neither in PHP-Scoper or in Box itself although I'm using it extensively and the codebase is not any smaller.

Meanwhile you can do `box compile --no-parallel` as an alternative. It is much slower but at least it won't cause that failure.

If Travis fails due to the long build time, you can do like I'm doing [here](https://github.com/humbug/php-scoper/blob/master/.travis.yml#L31) with PHP-Scoper.

**Pull request merged**